src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -890,36 +884,25 @@ internal override bool GetIsValueType(TypeSymbol typeSymbol, ConsList<TypeParame
 
             internal override TypeSymbolWithAnnotations WithModifiers(TypeSymbolWithAnnotations type, ImmutableArray<CustomModifier> customModifiers)
             {
-                return CreateNonLazyType(type._defaultType, type.NonNullTypesContext, type.IsAnnotated,
-                    type._treatPossiblyNullableReferenceTypeTypeParameterAsNullable, customModifiers);
-            }
-
-            internal override TypeSymbolWithAnnotations WithNonNullTypesContext(TypeSymbolWithAnnotations type, INonNullTypesContext nonNullTypesContext)
-            {
-                Debug.Assert(nonNullTypesContext != null);
-                return CreateNonLazyType(type._defaultType, nonNullTypesContext, type.IsAnnotated,
-                    type._treatPossiblyNullableReferenceTypeTypeParameterAsNullable, _customModifiers);
+                return CreateNonLazyType(type._defaultType, type.NullableAnnotation, customModifiers);
             }
 
             internal override TypeSymbol AsTypeSymbolOnly(TypeSymbol typeSymbol) => typeSymbol;
 
             internal override TypeSymbolWithAnnotations WithTypeAndModifiers(TypeSymbolWithAnnotations type, TypeSymbol typeSymbol, ImmutableArray<CustomModifier> customModifiers)
             {
-                return CreateNonLazyType(typeSymbol, type.NonNullTypesContext, isAnnotated: type.IsAnnotated,
-                    type._treatPossiblyNullableReferenceTypeTypeParameterAsNullable, customModifiers);
+                return CreateNonLazyType(typeSymbol, type.NullableAnnotation, customModifiers);
             }
 
-            internal override TypeSymbolWithAnnotations AsNullableReferenceType(TypeSymbolWithAnnotations type)
+            internal override TypeSymbolWithAnnotations AsNullableReferenceType(TypeSymbolWithAnnotations type, bool fromDeclaration)
             {
-                return CreateNonLazyType(type._defaultType, type.NonNullTypesContext, isAnnotated: true,
-                    type._treatPossiblyNullableReferenceTypeTypeParameterAsNullable, _customModifiers);
+                return CreateNonLazyType(type._defaultType, fromDeclaration ? NullableAnnotation.Nullable : NullableAnnotation.NullableBasedOnAnalysis, _customModifiers);
             }
 
             internal override TypeSymbolWithAnnotations AsNotNullableReferenceType(TypeSymbolWithAnnotations type)
```

Comment:
>AsNotNullableReferenceType [](start = 56, length = 26)

Should we have a "from declaration" flag here too? It looks like `AsNotNullableReferenceType` is used in method type inference, which is run during both binding and flow analysis.

Reply:
The client would be method type inference. Since it can be run during flow analysis, it feels like it should be passing a `fromDeclaration: false` in that scenario.

---
In reply to: [230951976](https://github.com/dotnet/roslyn/pull/30913#discussion_r230951976) [](ancestors = 230951976,230525559)

Reply:
> The client would be method type inference. Since it can be run during flow analysis, it feels like it should be passing a fromDeclaration: false in that scenario.

Feel free to come up with a scenario that would indicate the need, so far we don't have any tests indicating that.

---
In reply to: [230958183](https://github.com/dotnet/roslyn/pull/30913#discussion_r230958183) [](ancestors = 230958183,230951976,230525559)

Reply:
It seems that the distinction between `NotNullable` and `NotNullableBasedOnAnalysis` currently does nothing. So if we're incorrectly producing one rather than the other, there would be no way to observe that.

This makes me think one of those two flags should be removed. Then the question of whether we're producing the correct flag here becomes mute ;-)


---
In reply to: [230958960](https://github.com/dotnet/roslyn/pull/30913#discussion_r230958960) [](ancestors = 230958960,230958183,230951976,230525559)

Reply:
> This makes me think one of those two flags should be removed. Then the question of whether we're producing the correct flag here becomes mute ;-)

At the moment, ```NotNullableBasedOnAnalysis``` affects behavior of SetPossiblyNullableReferenceTypeTypeParametersAsNullable in the way that we don't overwrite types that flow from earlier analysis. More real usage is coming in the next commit.  

---
In reply to: [230959880](https://github.com/dotnet/roslyn/pull/30913#discussion_r230959880) [](ancestors = 230959880,230958960,230958183,230951976,230525559)

Reply:
I'm sorry, I didn't follow. I looked at all the references to `NotNullableBasedOnAnalysis` in this PR, and none of them behaves differently when encountering a `NotNullableBasedOnAnalysis` or a `NotNullable`.

---
In reply to: [230961770](https://github.com/dotnet/roslyn/pull/30913#discussion_r230961770) [](ancestors = 230961770,230959880,230958960,230958183,230951976,230525559)

Reply:
> I looked at all the references to NotNullableBasedOnAnalysis in this PR, and none of them behaves differently when encountering a NotNullableBasedOnAnalysis or a NotNullable.

The following method behaves differently:
```
        internal bool IsPossiblyNullableReferenceTypeTypeParameter()
        {
            return NullableAnnotation == NullableAnnotation.NotNullable && TypeSymbol.IsPossiblyNullableReferenceTypeTypeParameter();
        }

```


---
In reply to: [230962617](https://github.com/dotnet/roslyn/pull/30913#discussion_r230962617) [](ancestors = 230962617,230961770,230959880,230958960,230958183,230951976,230525559)